tests/unittests/MLTest.cpp

```diff
@@ -983,6 +983,204 @@ TEST_P(InterpreterAndCPU, testFindPixelRegression) {
+//              --+
+//      | a b c | |
+// M1 = | d e f | V
+//      | g h i |
+// =>
+//      | d a b |
+// M2 = | g e c |
+//      | h i f |
+static void generateMatrixRotationRecognitionData(Tensor &matricesA,
+                                                  Tensor &matricesB,
+                                                  Tensor &expected,
+                                                  PseudoRNG &PRNG) {
+
+  using CellIdx = std::pair<uint8_t, uint8_t>;
+  // List the indices in a clockwise ordering starting from the top left
+  // corner.
+  // Note: This does not include the cell in the middle given it is
+  // never rotated.
+  static const CellIdx clockwiseOrder[] = {{0, 0}, {0, 1}, {0, 2}, {1, 2},
+                                           {2, 2}, {2, 1}, {2, 0}, {1, 0}};
+  static const uint8_t possibleTargetCells =
+      sizeof(clockwiseOrder) / sizeof(clockwiseOrder[0]);
+  const unsigned numSamples = matricesA.dims()[0];
+  assert(expected.dims()[0] == numSamples &&
+         matricesB.dims()[0] == numSamples &&
+         "Size of the tensors is incompatible");
+  auto handleMatricesA = matricesA.getHandle<float>();
+  auto handleMatricesB = matricesB.getHandle<float>();
+  auto handleExpected = expected.getHandle<size_t>();
+
+  handleMatricesA.randomize(0.0, 2.0, PRNG);
+  handleMatricesB.randomize(0.0, 2.0, PRNG);
+  for (unsigned idx = 0; idx < numSamples; ++idx) {
+    // Narrow the search space for the network by using only 0s and 1s.
+    for (size_t rowIdx = 0; rowIdx != 3; ++rowIdx) {
+      for (size_t columnIdx = 0; columnIdx != 3; ++columnIdx) {
+        handleMatricesA.at({idx, rowIdx, columnIdx}) =
+          handleMatricesA.at({idx, rowIdx, columnIdx}) > 1.0;
+        handleMatricesB.at({idx, rowIdx, columnIdx}) =
+          handleMatricesB.at({idx, rowIdx, columnIdx}) > 1.0;
+      }
+    }
+    // Toss a coin and create a rotation relationship or not.
+    bool shouldHaveRotation = PRNG.nextRandInt(0, 1);
+    handleExpected.at({idx, 0}) = shouldHaveRotation;
+    if (shouldHaveRotation) {
+      // On a 3x3 matrix we have 8 different possbile clockwise steps.
+      // Pick one.
+      size_t clockwiseSteps = PRNG.nextRandInt(0, possibleTargetCells - 1);
+      // Generate the rotation matrix from A.
+      // The center never changes.
+      handleMatricesB.at({idx, 1, 1}) = handleMatricesA.at({idx, 1, 1});
+      // Fetch the cell registered in the clockwiseOrder at the desired step.
+      for (size_t i = 0; i != possibleTargetCells; ++i) {
+        const CellIdx &sourceCellIdx = clockwiseOrder[i];
+        const CellIdx &targetCellIdx =
+            clockwiseOrder[(i + clockwiseSteps) % possibleTargetCells];
+        handleMatricesB.at({idx, targetCellIdx.first, targetCellIdx.second}) =
+            handleMatricesA.at({idx, sourceCellIdx.first, sourceCellIdx.second});
+      }
+    } else {
+      // There is a high probability that A and B don't have a rotation
```

Comment:
I am not sure why we need this part. Even if our matrices support only boolean values, the probability of randomly selecting a matrix that happen to be a rotation of another matrix is very low.  (2^9 / 8), which is way below the accuracy that we expect from a network (90% accuracy is fine). If we want to be extra conservative we can select the input values -1, 0, 1, and this will reduce the probability of accidentally generating a random matrix significantly. The neural network should learn to generalize and ignore these inputs. In fact, introducing noise is a way to make neural networks to generalize. So, I don't think that accidentally generating the wrong label in a very small number of cases is something that we should worry about.